### ISSUE TITLE
Incorrect KDBX 4 hashed blocks HMAC

### ISSUE THREAD

**Issue by Footmau** [tobischo/gokeepasslib#37]
Original KeePass application can not open KDBX 4 database after it was modified using gokeepasslib because of incorrect HMAC hash inside first (and probably other) block.

I tried to investigate problem to make  pull request but I'm really new in Go and can not figure out how to fix it.

**Comment by tobischo** [comment]
Thank you for reporting this issue.

Would you be able to provide a description of the steps you took in a bit more detail as well as a code sample which can be used to recreate the issue?

Otherwise I am afraid I will have as hard a time as you in trying to figure out where the issue is coming from and how to fix it.

**Comment by Footmau** [comment]
This is [problem test db](https://drive.google.com/file/d/19qYtQ6hyVmDvCOqz58mqZnNxANvpGMRv/view?usp=sharing) (password "123").

Steps to reproduce:
1. Open db in original KeePass to ensure that it opens
2. Made any modification using gokeepasslib (for example rename entry)
3. Try to open it again in KeePass (gokeepasslib opens well, it doesnt check HMAC of every block)

I can not provide complete code, because I'm using gokeepasslib in the android app and I implemented facade to access it through JNI. This is how I save the db
```
func SaveDatabase() []byte {
	tmpDB := gokeepasslib.NewDatabase()
	deepcopy.Copy(tmpDB, db)
	var data bytes.Buffer
	w := bufio.NewWriter(&data)

	tmpDB.LockProtectedEntries()
	keepassEncoder := gokeepasslib.NewEncoder(w)

	if err := keepassEncoder.Encode(tmpDB); err != nil {
		return nil
	}

	w.Flush()
	return data.Bytes()
}
```

Also I can try to implement example code to reproduce the issue if it helps. But I think the problem will only appear on this specific database.

**Comment by tobischo** [comment]
Best as I can tell the problem of the invalid HMAC calculation comes from not properly following the documentation (not really a surprise 🙈)
https://github.com/tobischo/gokeepasslib/blob/master/blocks.go#L104

See `Improved Data Authentication` at https://keepass.info/help/kb/kdbx_4.html
> The HMAC is computed over i ‖ n ‖ C 
Right now it is only calculated over `C`

While I do not have a windows computer at hand right now, I am thinking that the fix for this problem should probably separate from adding the HMAC check here.
That way anyone who has used the faulty implementation so far to write KDBX4 files can read and write the files using an intermediate version and then upgrade to the version that does check for the HMAC.

**Comment by Footmau** [comment]
Yes, this way is better for me too.
Thanks for answer. Will you fix it?

**Comment by tobischo** [comment]
I am planning to.
I cannot make promises as to when I will get it done though.
While I am reasonable certain that I know where I have to make changes, I am not 100% certain as to how exactly they have to look like.
While I understand the documentation of the file format, it is not very... precise.
I'd need to read up more on the corresponding C# code and understand the crypto methods there more to get it working properly.

I can recreate the issue though, so it should be doable

**Comment by tobischo** [comment]
Right now I am looking into other implementations of this, which I had hoped would be easier to grok for me than C#, e.g. a python implementation: https://github.com/pschmitt/pykeepass/blob/b88f4dccd4430aa45e928f478d707ed2efd48f80/pykeepass/kdbx_parsing/kdbx4.py#L156

Currently I am at
```go
// ...
keyBuilder := sha512.New()
keyBuilder.Write(masterSeed)
keyBuilder.Write(transformedKey)
keyBuilder.Write([]byte{0x01})
hashKey := keyBuilder.Sum(nil)

blockKeyBuilder := sha512.New()
binary.Write(blockKeyBuilder, binary.LittleEndian, index)
blockKeyBuilder.Write(hashKey)
blockKey := blockKeyBuilder.Sum(nil)

mac := hmac.New(sha256.New, blockKey)
// Write index
binary.Write(mac, binary.LittleEndian, index)
// Write length
binary.Write(mac, binary.LittleEndian, length)
// Write data
mac.Write(data)
expectedHash := mac.Sum(nil)
// ...
```
something is still wrong though and I am not entirely sure what that is

**Comment by Footmau** [comment]
I tried to investigate debugging the decode methid in KeePass C# project and noted that for KDBX 4 used another key.
```
private void ComputeKeys(out byte[] pbCipherKey, int cbCipherKey,
	out byte[] pbHmacKey64)
{
	byte[] pbCmp = new byte[32 + 32 + 1];
	try
	{
		Debug.Assert(m_pbMasterSeed != null);
		if(m_pbMasterSeed == null)
			throw new ArgumentNullException("m_pbMasterSeed");
		Debug.Assert(m_pbMasterSeed.Length == 32);
		if(m_pbMasterSeed.Length != 32)
			throw new FormatException(KLRes.MasterSeedLengthInvalid);
		Array.Copy(m_pbMasterSeed, 0, pbCmp, 0, 32);

		Debug.Assert(m_pwDatabase != null);
		Debug.Assert(m_pwDatabase.MasterKey != null);
		ProtectedBinary pbinUser = m_pwDatabase.MasterKey.GenerateKey32Ex(
			m_pwDatabase.KdfParameters, m_slLogger);
		Debug.Assert(pbinUser != null);
		if(pbinUser == null)
			throw new SecurityException(KLRes.InvalidCompositeKey);
		byte[] pUserKey32 = pbinUser.ReadData();
		if((pUserKey32 == null) || (pUserKey32.Length != 32))
			throw new SecurityException(KLRes.InvalidCompositeKey);
		Array.Copy(pUserKey32, 0, pbCmp, 32, 32);
		MemUtil.ZeroByteArray(pUserKey32);

		pbCipherKey = CryptoUtil.ResizeKey(pbCmp, 0, 64, cbCipherKey);

		pbCmp[64] = 1;
		using(SHA512Managed h = new SHA512Managed())
		{
			pbHmacKey64 = h.ComputeHash(pbCmp);
		}
	}
	finally { MemUtil.ZeroByteArray(pbCmp); }
}
```
_pbCipherKey_ for KDBX 3 and _pbHmacKey64_ for KDBX 4. This is out parameters of the method (KeePass-2.42.1-Source\KeePassLib\Serialization\KdbxFile.cs).

So maybe we should use something instead of _transformedKey_.

**Comment by tobischo** [comment]
That's already covered with
```go
keyBuilder := sha512.New()
keyBuilder.Write(masterSeed)
keyBuilder.Write(transformedKey)
keyBuilder.Write([]byte{0x01})
hashKey := keyBuilder.Sum(nil)
```
where in fact just using `transformedKey` is not enough. It has to be used together with `masterSeed` and the additional `1 byte`

**Comment by tobischo** [comment]
ok, I think I got it now after comparing step by step with the python implementation.
I'll still have to make the code pretty though

**Comment by tobischo** [comment]
Please take a look at my changes from #38 and see if that works for you

**Comment by Footmau** [comment]
Seems it works fine for me.
Thanks a lot for quick fix!
It could be useful also to add optional HMAC check while decoding.

**Comment by tobischo** [comment]
I'll make it mandatory, using the same logic, however that will happen in another commit

**Comment by tobischo** [comment]
This fix was released in v2.0.2

**Comment by tobischo** [comment]
The version with included hmac check was released in v2.0.3
